src/heartbeats.cpp

```diff
@@ -452,330 +299,120 @@ MegaBackupMonitor::MegaBackupMonitor(MegaClient *client)
-                                           ));
+                                           info.state(),
+                                           info.subState(),
+                                           info.extra().c_str(),
+                                           nullptr));
 }
 
+#ifdef ENABLE_SYNC
 
-void MegaBackupMonitor::registerBackupInfo(const MegaBackupInfo &info)
+void MegaBackupMonitor::registerBackupInfo(const MegaBackupInfo &info, UnifiedSync* syncPtr)
 {
     string localFolderEncrypted(mClient->cypherTLVTextWithMasterKey("lf", info.localFolder()) );
     string deviceIdHash = mClient->getDeviceidHash();
 
     mClient->reqs.add(new CommandBackupPut(mClient, info.type(), info.backupName(), info.megaHandle(),
                                            localFolderEncrypted.c_str(),
                                            deviceIdHash.c_str(),
-                                           info.state(), info.subState(), info.extra().c_str()
-                                           ));
+                                           info.state(), info.subState(), info.extra().c_str(),
+                                           [this, syncPtr](Error e, handle h){ if (!e) digestPutResult(h, syncPtr); }));
 }
 
-#ifdef ENABLE_SYNC
 
-void MegaBackupMonitor::updateOrRegisterSync(MegaSync *sync)
+void MegaBackupMonitor::updateOrRegisterSync(UnifiedSync& us)
 {
-    if (!sync)
-    {
-        return;
-    }
-    int syncTag = sync->getTag();
-    auto config = mClient->syncConfigs->get(syncTag);
-    handle backupId = config ? config->getBackupId() : UNDEF;
-
-    std::unique_ptr<MegaBackupInfo> info = ::mega::make_unique<MegaBackupInfoSync>(mClient, *sync, backupId);
+    MegaBackupInfoSync currentInfo(us);
 
-    if (info->backupId() == UNDEF) // not registered (or pending registration)
+    if (!us.mBackupInfo && us.mConfig.getBackupId() == UNDEF) // not registered yet
     {
-        if (mPendingSyncPuts.find(syncTag) == mPendingSyncPuts.end()) //new backup, register required
-        {
-            registerBackupInfo(*info.get());
-
-            // queue callback to process the backupId when received
-            mPendingSyncPuts.insert(syncTag);
-            mPendingBackupPutCallbacks.push_back([this, syncTag](handle backupId)
-            {
-                onSyncBackupRegistered(syncTag, backupId);
-            });
-        }
-        else // registration in-flight: backupId not received yet, let's queue the update
-        {
-            LOG_debug << " Queuing sync update, register is on progress for sync: " << info->localFolder();
-            mPendingSyncUpdates[syncTag].reset(info.release()); // we replace any previous pending updates
-        }
+        us.mBackupInfo = ::mega::make_unique<MegaBackupInfoSync>(us);
+        registerBackupInfo(currentInfo, &us);
     }
-    else //update
+    else if (us.mConfig.getBackupId() != UNDEF  &&
```

Comment:
Yes, I totally agree - we need the backupId changes.   I've left it like this on the assumption that initially we create a backup with id UNDEF, if it' still UNDEF then we try to register it, once that succeeds we set the Id and after that the normal heartbeats can be sent.     But all that can be addressed by the upcoming work to use backupId instead of `tag`.

Reply:
Well, there was some code in place from Pablo that aimed to keep those updates in a queue, so once the backup gets registered, the updates got sent out. However, with this branch, those changes are lost.
@polmr probably that's not needed anymore with the pre-registration, since the user won't be able to change the backup params until it's registered in the API, right? But something to take into account for the registration rework.

Reply:
Yep, we should remove the `Register` from this `updateOrRegisterSync` and this will always be updates and backupId will be always != UNDEF if there exists one unifiedSync. No need to control updates in between and all.